# DS-SF-36 | 01 | What is Data Science | Assignment | Starter Code

## Python

> ## "Data scientists' most basic, universal skill is the ability to write code."
>
> Harvard Business Review | “Data Scientists: The Sexiest Job of the 21st Century” (2012)
>
> (https://hbr.org/2012/10/data-scientist-the-sexiest-job-of-the-21st-century)

Today, to give you some more practice beyond the course pre-requisites, we are going to implement a few functions in Python.  This assignment covers to some extend the following topics:

- Functions (defining and using your own functions but also calling functions from packages)
- Loops
- Arithmetic operations
- Conditional statements

**Don't worry if you get stuck.  Ask around, review the answer key, and ask around more.  As this course progresses, your programming proficiency will increase.**

> ### Question 1.  Multiples of 3 and 5
>
> If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9.  The sum of these multiples is 23.
>
> Find the sum of all the multiples of 3 or 5 below 1,000.
>
> (Source: [Project Euler | Problem 1](https://projecteuler.net/problem=1))

In [9]:
sum =0;
for i in range(3,1000):
    if(i%3==0 or i%5==0):
       sum+=i;
print(sum);


233168


Answer: 233168

> ### Question 2.  Estimating square roots
>
> Given a real number $m$, let's define the series $u$ as follow:
> - $u_0 = 1$
> - $u_{n+1} = \frac{u_n ^ 2 + m}{2u_n}$
>
>
> Implement the calculations of the series $u$ above to estimate square roots.  Verify that $\sqrt{144} = 12$ and use your function to calculate $\sqrt{1024}$.

In [1]:
arr = dict([(0,1)])
errorRate = 0.1;

def computeDiff(a,b):
    return abs(a-b);

def computeRoot(m):    
    for i in range(1,m+1):
        a = arr[i-1];
        root = (a*a+m)/(2*a);
        arr.update({i:root});
        if computeDiff(arr[i],arr[i-1])* 100 <= errorRate:            
            return arr[i];
    
    return arr[m];
 
m = int(input('Enter your number: '));

tot =computeRoot(m);
print(tot);

Enter your number: 12365
111


Answer: 32

> ### Question 3.  Prime Numbers
>
> A prime (number) is a natural number greater than 1 that has no positive divisors other than 1 and itself.  ([Wikipedia](https://en.wikipedia.org/wiki/Prime_number))
>
> Calculate all primes below 1,000.  What's their sum?

In [13]:
arr =[]


def getPrimeSeq(n):
    for i in range(1,n):
        chk=1;
        for j in range(2,int(i/2)+1):
            if(i%j==0):
                chk=0;
                break;
        if(chk):
            arr.append(i);
n=1000;
getPrimeSeq(n);
# print(arr);
print(sum(arr))


76128


Answer: 76128

> ### Question 4.  Largest prime factor
>
> The prime factors of 13195 are 5, 7, 13 and 29.
>
> What is the largest prime factor of the number 600851475143?
>
> (Source: [Project Euler | Problem 3](https://projecteuler.net/problem=3))

In [14]:
num = 600851475143  
i = 2  
while i * i < num:  
     while num % i == 0:  
         num = num / i  
     i = i + 1  
print(num) 

6857


Answer: 6857

> ### Question 5.  Mean
>
> Write a function to calculate the mean (average) of a list.
>
> What's the mean of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [15]:
arr=[10,8,13,9,11,14,6,4,12,7,5]

print(sum(arr)/len(arr))


9


Answer: 9

> ### Question 6.  Sample standard deviation
>
> Write a function to calculate the standard deviation of a sample.
>
> Given the sample $x_1, x_2, ..., x_N$, its standard deviation is defined as $s = \sqrt{\frac{1}{N - 1} \sum_{i = 1}^{N} (x_i  - \bar{x})^2}$, with $\bar{x}$ as the sample mean.
>
> What's the standard deviation of the following sample: 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?
>
> ([Wikipedia](https://en.wikipedia.org/wiki/Standard_deviation#Sample_standard_deviation))

In [44]:
data=[10,8,13,9,11,14,6,4,12,7,5];

s=0;

for i in data:
    s +=i;
n=len(data)

avg = s/n;

tot =0;
for i in data:
    tot +=pow(i-avg,2);

temp = tot/(len(data)-1);
result = temp**(0.5)

print(result);


3.31662479036


Answer: 3.31662479036

> ### Question 7.  Median
>
> Write a function to calculate the median ("middle value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Median))
>
> What's the median of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [31]:
data=[10,8,13,9,11,14,6,4,12,7,5]

data = sorted(data)
n = len(data);
mid = 0.0;
if n%2 == 1:
    i = n/2;
    mid=data[i];
else:
    x=n/2;
    y=x+1;
    mid=(data[x]+data[y])/2;

print(mid)

9


Answer: 9

> ### Question 8.  Mode
>
> Write a function to calculate the mode ("most frequent value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Mode_(statistics)))
>
> What's the mode of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7 and 5?  How about the mode of 8, 8, 8, 8, 8, 8, 19, 8, 8 and 8?

In [32]:
#data=[10,8,13,9,11,14,6,4,12,7,5]

data=[8, 8, 8, 8, 8, 8, 19, 8, 8 ,8 ]

data = sorted(data)
n= len(data)
maxCount =0;
count =0;
mode =0;
for i in range(1,n):
    if data[i] == data[i-1]:
        count +=1;
    else:
        if maxCount < count:
            maxCount=count;
            mode=data[i-1];
        count =0;

if mode > 0:
    print(mode)
else:
    print('mode value does not exist')

8


Answer: 
1. mode value does not exist
2. 8